WcaOnRails/app/models/upload_json.rb

```diff
@@ -0,0 +1,127 @@
+    if valid?
+      competition = Competition.find(competition_id)
+      persons_to_import = []
+      parsed_json["persons"].each do |p|
+        new_person_attributes = p.merge(competitionId: competition_id)
+        persons_to_import << InboxPerson.new(new_person_attributes)
+      end
+      results_to_import = []
+      scrambles_to_import = []
+      parsed_json["events"].each do |event|
+        event["rounds"].each do |round|
+          # Import results for round
+          round["results"].each do |result|
+            individual_results = result["results"]
+            # Pad the results with 0 up to 5 results
+            individual_results.fill(0, individual_results.length...5)
+            new_result_attributes = {
+              personId: result["personId"],
+              pos: result["position"],
+              eventId: event["eventId"],
+              roundTypeId: round["roundId"],
+              formatId: round["formatId"],
+              best: result["best"],
+              average: result["average"],
+              value1: individual_results[0],
+              value2: individual_results[1],
+              value3: individual_results[2],
+              value4: individual_results[3],
+              value5: individual_results[4],
+            }
+            new_res = InboxResult.new(new_result_attributes)
+            # Using this way of setting the attribute saves one SELECT per result
+            # to validate the competition presence.
+            # (a lot of time considering all the results to import!)
+            new_res.competition = competition
+            results_to_import << new_res
+          end
+
+          # Import scrambles for round
+          round["groups"].each do |group|
+            ["scrambles", "extraScrambles"].each do |scramble_type|
+              group[scramble_type]&.each_with_index do |scramble, index|
+                new_scramble_attributes = {
+                  competitionId: competition_id,
+                  eventId: event["eventId"],
+                  roundTypeId: round["roundId"],
+                  groupId: group["group"],
+                  isExtra: scramble_type == "extraScrambles",
+                  scrambleNum: index+1,
+                  scramble: scramble,
+                }
+                scrambles_to_import << Scramble.new(new_scramble_attributes)
+              end
+            end
+          end
+        end
+      end
+      begin
+        ActiveRecord::Base.transaction do
+          InboxPerson.where(competitionId: competition_id).delete_all
+          InboxResult.where(competitionId: competition_id).delete_all
+          Scramble.where(competitionId: competition_id).delete_all
+          InboxPerson.import!(persons_to_import)
+          Scramble.import!(scrambles_to_import)
+          InboxResult.import!(results_to_import)
+        end
+        true
+      rescue ActiveRecord::RecordNotUnique
+        errors.add(:results_file, "Duplicate personId in JSON.")
```

Comment:
I'd worry we could get `ActiveRecord::RecordNotUnique` for other reasons in the future. Is it possible to turn this into an `ActiveRecord::RecordInvalid` by adding a validation to a model somewhere? That feels a little more future proof to me, although it might come with a performance hit?

Reply:
Hum I tried it and that's not trivial.
The way we import data is the following:
  - we delete all `InboxPerson` for the given competition
  - we bulk import all the new `InboxPerson` (which is done in one single insert statement)

Because of this, validating the `InboxPerson` for uniqueness of `(id, competitionId)` because there is no record saved to the db at the time of the validation.

Do you have any good idea to approach this?
(beside the obvious "don't do bulk insert" :p)